xmppserver/src/main/java/org/jivesoftware/admin/servlet/BlowfishMigrationServlet.java

```diff
@@ -0,0 +1,440 @@
+     * @throws IllegalStateException if already using PBKDF2 or not using Blowfish
+     * @throws RuntimeException if migration fails (with rollback)
+     */
+    private MigrationResult migrateBlowfishToPBKDF2() throws Exception {
+        // 1. Verify preconditions
+        String currentKdf = JiveGlobals.getBlowfishKdf();
+        if (JiveGlobals.BLOWFISH_KDF_PBKDF2.equalsIgnoreCase(currentKdf)) {
+            throw new IllegalStateException("Already using PBKDF2 - no migration needed");
+        }
+
+        String encryptionAlgorithm = JiveGlobals.getEncryptionAlgorithm();
+        if (!ENCRYPTION_ALGORITHM_BLOWFISH.equalsIgnoreCase(encryptionAlgorithm)) {
+            throw new IllegalStateException("Encryption algorithm is " + encryptionAlgorithm +
+                    ", not Blowfish. Migration only applies to Blowfish.");
+        }
+
+        Log.info("Starting Blowfish migration from SHA1 to PBKDF2...");
+
+        // 2. Migrate XML properties FIRST (fail fast before database changes)
+        // This migrates encrypted properties in openfire.xml (e.g., database credentials)
+        int xmlMigrated = 0;
+        try {
+            xmlMigrated = JiveGlobals.migrateXMLPropertiesFromSHA1ToPBKDF2();
+            Log.info("Migrated {} XML properties from openfire.xml", xmlMigrated);
+        } catch (Exception e) {
+            // XML migration failed - stop before making any database changes
+            throw new RuntimeException("XML property migration failed. " +
+                    "No database changes were made. " + e.getMessage(), e);
+        }
+
+        // 3. Find all Blowfish-encrypted properties in database
+        List<EncryptedProperty> dbProperties = getAllEncryptedPropertiesFromDatabase();
+        Log.info("Found {} encrypted properties in database to migrate", dbProperties.size());
+
+        // 4. Get the master encryption key from JiveGlobals
+        String masterKey = JiveGlobals.getMasterEncryptionKey();
+        if (masterKey == null) {
+            Log.info("No custom encryption key configured - using default Blowfish key for migration");
+        }
+
+        // 5. Initialise both SHA1 and PBKDF2 encryptors
+        // We need two separate instances with different KDFs:
+        // - SHA1: for decrypting existing properties
+        // - PBKDF2: for re-encrypting with new KDF
+        Blowfish sha1Blowfish = new Blowfish();
+        sha1Blowfish.setKey(masterKey, JiveGlobals.BLOWFISH_KDF_SHA1);
+
+        Blowfish pbkdf2Blowfish = new Blowfish();
+        pbkdf2Blowfish.setKey(masterKey, JiveGlobals.BLOWFISH_KDF_PBKDF2);
+
+        // 6. Migrate each database property within a transaction
+        Connection con = null;
+        boolean originalAutoCommit = true;
+        int migrated = 0;
+        int failed = 0;
+        List<String> failedProperties = new ArrayList<>();
+
+        try {
+            con = DbConnectionManager.getConnection();
```

Comment:
This should possibly be replaced by `org.jivesoftware.database.DbConnectionManager#getTransactionConnection`